File: src/engine/repoSelector.ts
// ============================================================
// Repo Selector — LLM-Driven Intent-Based Repo Selection
// Understands user intent → picks BEST repos → builds product
// ============================================================

import { generate, generateJSON, embed, cosineSimilarity } from "@/llm/provider";
import { CapabilityCategory } from "@/lib/agents/types";

export interface UserIntent {
  domain: string;
  features: string[];
  techPreferences: string[];
  capabilities: CapabilityCategory[];
  summary: string;
  requiredRoles?: string[]; // Added for targeted retries
}

export interface RankedRepo {
  name: string;
  description: string;
  score: number;
  reason: string;
  role: CapabilityCategory | "general" | string;
  confidence: number;
}

/**
 * Step 1: Extract structured intent from user input
 */
export async function extractIntent(userInput: string): Promise<UserIntent> {
  const prompt = `Extract structured intent from this product idea:

"${userInput}"

Return JSON with:
{
  "domain": "the primary domain (e.g., AI, DevTools, Healthcare)",
  "features": ["list of key features needed"],
  "techPreferences": ["preferred technologies mentioned"],
  "capabilities": ["which of these are needed: memory, agent, rag, ui, automation, model-serving, data, security, infra, communication"],
  "summary": "one sentence summary of what the user wants to build"
}`;

  try {
    const intent = await generateJSON<UserIntent>(prompt, "You are a product analysis AI. Be precise and actionable.");
    return intent;
  } catch {
    // Fallback: keyword-based intent extraction
    return extractIntentLocal(userInput);
  }
}

/**
 * Step 2: Rank repos based on intent using LLM
 */
export async function rankRepos(
  repos: Array<{ name: string; description: string | null; stars: number; language: string | null; topics: string[] }>,
  intent: UserIntent
): Promise<RankedRepo[]> {
  const roleToQueryMap: Record<string, string[]> = {
    'workflow': ['n8n', 'temporal', 'airflow', 'zapier'],
    'execution': ['playwright', 'selenium', 'puppeteer', 'webdriver'],
    'agent': ['autogen', 'langchain', 'llamaindex', 'crewai'],
    'orchestration': ['langgraph']
  };

  const targetedQueries: string[] = [];
  if (intent.requiredRoles) {
    intent.requiredRoles.forEach(role => {
      if (roleToQueryMap[role]) {
        targetedQueries.push(...roleToQueryMap[role]);
      }
    });
  }

  const prompt = `You are selecting the BEST repositories to build a product.

User Intent:
${JSON.stringify(intent, null, 2)}

Targeted Overrides (Must strongly prioritize if present): ${targetedQueries.join(', ')}

Available Repositories:
${repos.map((r, i) => `${i + 1}. ${r.name}: ${r.description || "No description"} (${r.stars} stars, ${r.language || "unknown"})`).join("\n")}

Return top 7 repos as JSON array:
[
  {
    "name": "repo full name",
    "description": "brief description",
    "score": 0-10,
    "reason": "why this repo fits the intent",
    "role": "memory | agent | rag | ui | automation | model-serving | data | security | infra | communication | general",
    "confidence": 0.0-1.0
  }
]`;

  try {
    const ranked = await generateJSON<RankedRepo[]>(prompt, "You are an expert repo selector AI. Be critical and selective.");
    return ranked.sort((a, b) => b.score - a.score);
  } catch {
    // Fallback: semantic similarity ranking
    return rankReposLocal(repos, intent, targetedQueries);
  }
}

/**
 * Full selector pipeline: Extract intent → Rank repos → Return top picks
 */
export async function selectBestRepos(
  userInput: string,
  repos: Array<{ name: string; description: string | null; stars: number; language: string | null; topics: string[] }>
): Promise<{ intent: UserIntent; rankedRepos: RankedRepo[] }> {
  const intent = await extractIntent(userInput);
  const rankedRepos = await rankRepos(repos, intent);
  return { intent, rankedRepos };
}

/**
 * Retries or runs targeted matching.
 */
export async function selectBestReposFromIntent(
  intent: UserIntent,
  repos: Array<{ name: string; description: string | null; stars: number; language: string | null; topics: string[] }>
): Promise<{ intent: UserIntent; rankedRepos: RankedRepo[] }> {
  const rankedRepos = await rankRepos(repos, intent);
  return { intent, rankedRepos };
}

// ============================================================
// Fallback: Local keyword-based extraction
// ============================================================

function extractIntentLocal(userInput: string): UserIntent {
  const lower = userInput.toLowerCase();
  const capabilities: CapabilityCategory[] = [];

  const capKeywords: Record<CapabilityCategory, string[]> = {
    memory: ["database", "storage", "vector", "memory", "cache", "redis", "db"],
    agent: ["agent", "autonomous", "copilot", "assistant", "bot", "workflow"],
    rag: ["search", "retrieval", "rag", "document", "knowledge", "semantic"],
    ui: ["dashboard", "interface", "frontend", "ui", "visualization", "react"],
    automation: ["automat", "schedule", "pipeline", "ci/cd", "trigger", "cron"],
    "model-serving": ["model", "llm", "gpt", "inference", "ai", "ml", "deploy"],
    data: ["data", "analytics", "etl", "stream", "metrics"],
    security: ["security", "auth", "encrypt", "privacy"],
    infra: ["docker", "kubernetes", "cloud", "deploy", "infrastructure"],
    communication: ["chat", "real-time", "message", "notification"],
  };

  for (const [cap, keywords] of Object.entries(capKeywords)) {
    if (keywords.some((kw) => lower.includes(kw))) {
      capabilities.push(cap as CapabilityCategory);
    }
  }

  if (capabilities.length === 0) capabilities.push("agent", "rag");

  return {
    domain: "Technology",
    features: userInput.split(" ").filter((w) => w.length > 4).slice(0, 5),
    techPreferences: [],
    capabilities,
    summary: userInput.slice(0, 200),
  };
}

async function rankReposLocal(
  repos: Array<{ name: string; description: string | null; stars: number; language: string | null; topics: string[] }>,
  intent: UserIntent,
  targetedQueries: string[] = []
): Promise<RankedRepo[]> {
  const intentText = `${intent.domain} ${intent.features.join(" ")} ${intent.capabilities.join(" ")} ${targetedQueries.join(" ")}`;

  const scored = repos.map((repo) => {
    const repoText = `${repo.name} ${repo.description || ""} ${repo.topics.join(" ")}`.toLowerCase();
    const intentLower = intentText.toLowerCase();

    // Simple keyword overlap scoring
    const intentWords = intentLower.split(/\s+/);
    const matchCount = intentWords.filter((w) => w.length > 2 && repoText.includes(w)).length;
    let keywordScore = Math.min(matchCount / Math.max(intentWords.length, 1) * 10, 10);
    
    // Targeted boost
    if (targetedQueries.some(q => repoText.includes(q.toLowerCase()))) {
      keywordScore += 5; // Heavy boost for explicitly targeting
    }

    // Star-based popularity score
    const starScore = Math.min(repo.stars / 10000, 3);

    // Combined score
    const totalScore = keywordScore * 0.7 + starScore * 0.3;

    // Determine role
    const role = determineRole(repoText);

    return {
      name: repo.name,
      description: repo.description || "",
      score: Math.round(totalScore * 10) / 10,
      reason: `Matches ${matchCount} intent keywords with ${repo.stars} stars`,
      role,
      confidence: Math.min(totalScore / 10, 1),
    } as RankedRepo;
  });

  return scored.sort((a, b) => b.score - a.score).slice(0, 7);
}

function determineRole(repoText: string): CapabilityCategory {
  const roleScores: Record<CapabilityCategory, number> = {
    memory: 0, agent: 0, rag: 0, ui: 0, automation: 0,
    "model-serving": 0, data: 0, security: 0, infra: 0, communication: 0,
  };

  const roleKeywords: Record<CapabilityCategory, string[]> = {
    memory: ["vector", "embed", "chroma", "redis", "database", "storage", "db"],
    agent: ["agent", "langchain", "langgraph", "crewai", "autogen", "orchestrat"],
    rag: ["rag", "retrieval", "search", "semantic", "llamaindex", "document"],
    ui: ["react", "vue", "frontend", "dashboard", "ui", "component"],
    automation: ["automation", "workflow", "pipeline", "n8n", "trigger"],
    "model-serving": ["inference", "serving", "deploy", "llm", "gpt", "model"],
    data: ["data", "analytics", "etl", "stream", "kafka"],
    security: ["security", "auth", "oauth", "encryption"],
    infra: ["docker", "kubernetes", "cloud", "terraform", "nginx"],
    communication: ["chat", "websocket", "real-time", "message"],
  };

  for (const [role, keywords] of Object.entries(roleKeywords)) {
    roleScores[role as CapabilityCategory] = keywords.filter((kw) => repoText.includes(kw)).length;
  }

  const bestRole = Object.entries(roleScores).sort(([, a], [, b]) => b - a)[0];
  return (bestRole[1] > 0 ? bestRole[0] : "agent") as CapabilityCategory;
}

File: python-backend/engine/repo_selector.py
"""
Repo Selector — LLM-driven intent extraction and intelligent repo ranking.
Replaces simple keyword matching with semantic understanding.
"""

from typing import Any, Optional
from llm.provider import LLMProvider, get_provider


async def extract_intent(
    user_input: str,
    provider: Optional[LLMProvider] = None,
) -> dict[str, Any]:
    """
    Extract structured intent from user's natural language input.

    Returns:
      - domain: The primary domain of the user's idea
      - required_capabilities: List of needed capability types
      - keywords: Search keywords for repo discovery
      - description: Expanded description of the intent
    """
    if provider is None:
        provider = get_provider()

    messages = [
        {
            "role": "system",
            "content": """You are an intent extraction engine for an AI product builder.
Given a user's natural language product idea, extract structured intent.

Return ONLY valid JSON:
{
  "domain": "<primary domain (e.g., 'AI Development Tools', 'Data Analytics', 'DevOps Automation')>",
  "required_capabilities": ["<capability1>", "<capability2>", ...],
  "keywords": ["<search_keyword1>", "<search_keyword2>", ...],
  "description": "<expanded 2-sentence description of what the user wants to build>"
}

Valid capability types: memory, agent, rag, ui, backend, automation""",
        },
        {
            "role": "user",
            "content": f"USER IDEA:\n{user_input}",
        },
    ]

    raw = await provider.chat(messages, temperature=0.3, max_tokens=400)
    data = provider.parse_json(raw)

    return {
        "domain": data.get("domain", "Technology"),
        "required_capabilities": data.get("required_capabilities", ["backend"]),
        "keywords": data.get("keywords", [user_input.split()[0] if user_input else "software"]),
        "description": data.get("description", user_input),
    }


async def rank_repos(
    repos: list[dict[str, Any]],
    intent: dict[str, Any],
    provider: Optional[LLMProvider] = None,
) -> list[dict[str, Any]]:
    """
    Rank repos by relevance to the extracted intent using LLM.

    Returns top repos with score, reason, and suggested role.
    """
    if provider is None:
        provider = get_provider()

    if not repos:
        return []

    repo_list = []
    for r in repos[:15]:  # Limit to top 15 for LLM context
        repo_list.append({
            "full_name": r.get("full_name", r.get("name", "")),
            "stars": r.get("stars", r.get("stargazers_count", 0)),
            "description": (r.get("description") or "")[:100],
            "language": r.get("language", ""),
            "topics": r.get("topics", []),
        })

    messages = [
        {
            "role": "system",
            "content": """You are a repo selection agent. Given a product intent and list of repos,
rank the top 7 most relevant repos. For each repo, provide:
- score: relevance score 0-1
- reason: one-line explanation of why it's relevant
- role: suggested role in the product (e.g., "Core AI Engine", "Data Storage", "API Layer")

Return ONLY valid JSON:
{
  "rankings": [
    {"full_name": "<owner/repo>", "score": <0-1>, "reason": "<why relevant>", "role": "<suggested role>"}
  ]
}""",
        },
        {
            "role": "user",
            "content": f"""INTENT:
Domain: {intent.get('domain', '')}
Required capabilities: {intent.get('required_capabilities', [])}
Description: {intent.get('description', '')}

REPOS:
{repo_list}""",
        },
    ]

    raw = await provider.chat(messages, temperature=0.3, max_tokens=800)
    data = provider.parse_json(raw)

    rankings = data.get("rankings", [])
    rank_map = {r["full_name"]: r for r in rankings if "full_name" in r}

    # Merge rankings into repos
    result = []
    for repo in repos:
        full_name = repo.get("full_name", repo.get("name", ""))
        if full_name in rank_map:
            rank = rank_map[full_name]
            result.append({
                **repo,
                "relevance_score": rank.get("score", 0.5),
                "selection_reasoning": rank.get("reason", ""),
                "suggested_role": rank.get("role", ""),
            })
        else:
            result.append({
                **repo,
                "relevance_score": 0.3,
                "selection_reasoning": "Not in top ranked repos",
                "suggested_role": "",
            })

    # Sort by relevance score
    result.sort(key=lambda r: r.get("relevance_score", 0), reverse=True)

    return result[:7]  # Return top 7


async def select_best_repos(
    user_input: str,
    repos: list[dict[str, Any]],
    provider: Optional[LLMProvider] = None,
) -> dict[str, Any]:
    """
    Full repo selection pipeline: extract intent → rank repos → select best.

    Returns:
      - intent: Extracted user intent
      - selected_repos: Top repos with scores and reasoning
    """
    if provider is None:
        provider = get_provider()

    # Step 1: Extract intent
    intent = await extract_intent(user_input, provider)
    print(f"[RepoSelector] Intent extracted: domain={intent['domain']}, "
          f"capabilities={intent['required_capabilities']}")

    # Step 2: Rank repos by intent
    ranked = await rank_repos(repos, intent, provider)
    print(f"[RepoSelector] Ranked {len(ranked)} repos")

    return {
        "intent": intent,
        "selected_repos": ranked,
    }
